### PR TITLE
Update logic for creating background tasks

### PR DESCRIPTION
**Pull request by alexaubry**
### Issues

When investigating crashes related to suspension of the application, we identified that we were creating a large number of extended-lifetime background tasks. It was observed that the application could be killed by the system because the limit of number of active tasks was not respected, and did not perform clean up upon expiration in time.

In the context of this API and PR, we define these terms as:

- Background Task: UIKit component for asking for extended lifetime when the app is moved to the background
- Background Activity: action our code performs inside a background task

### Causes

We currently create background activities for many reasons. `BackgroundActivityFactory` is the component that starts and ends them. Every time it receives a request to begin an activity, it asks the `UIApplication` to begin a new background task.

This approach is fragile, as this increases the number of active tasks, bringing us closer to the limit. Furthermore, the clients of `BackgroundActivity` are responsible for ending the task on `UIApplication` – which is error prone. If the client does not call the method to end the activity returned by the factory, the app will crash.

In addition, it was easy to ignore the expiration handler provided by `UIApplication` and skip the cleanup part, which is also a cause of termination by the application watchdog.

### Solutions

As suggested in our communications with the Apple Developer support, we refactored the `BackgroundActivityFactory` to be the only object that starts and ends background tasks. It can only have one active UIKit background task at a time.

*TL;DR: we move the responsibility of creating UIKit tasks and handling expiration away from the activities.*

When you request a new background activity, the factory checks whether there is an active UIKit background task. If no, it starts one and returns a token object for the activity. If yes, it increases the number of active activities, without starting a new UIKit background task. If it returns `nil`, it means that the application is being terminated and that the activity cannot be done safely.

When you mark the activity as ended, the factory decreases the count of active activities. If the count reaches 0, it marks its underlying UIKit task as ended.

It is also responsible for observing expiration. When it creates the UIKit task, it passes a block that will call all the activities expiration handlers, and end the UIKit task regardless of what the activities do in their expiration handler, to make sure the app doesn't crash.

We also block the creation of new background tasks when the app is being expired to avoid more crashes.

**Testing**

To test these changes, we create a mock of the `UIApplication` background task API that can simulate limits and the expiration of the app.

### Dependencies

Needs releases with:

- [x] https://github.com/wireapp/wire-ios-utilities/pull/68

### Notes

We need to update the packages dependent on system before we can merge this pull request.

### CODE REVIEW

File: Source/Background/BackgroundActivityFactory.swift [hunk 1]
```diff
@@ -19,28 +19,150 @@
...
+            activities.removeAll()
+
+            finishBackgroundTask()
+            currentBackgroundTask = UIBackgroundTaskInvalid
+        }
+    }
+
+    /// Ends the current background task.
+    func finishBackgroundTask() {
+        if let currentBackgroundTask = self.currentBackgroundTask {
```
Comment (typfel) [inline Source/Background/BackgroundActivityFactory.swift]: I guess this must also be called on the isolation queue? Maybe make this method private and add another public method which wraps it in the isolation queue?

File: Source/Background/BackgroundActivityFactory.swift [hunk 2]
```diff
@@ -19,28 +19,150 @@
...
+                finishBackgroundTask()
+            }
+        }
+    }
+
+    // MARK: - Helpers
+
+    /// Starts the background activity of the system allows it.
+    func startActivityIfPossible(_ name: String, _ expirationHandler: (() -> Void)?) -> BackgroundActivity? {
+        return isolationQueue.sync {
```
Comment (typfel) [inline Source/Background/BackgroundActivityFactory.swift]: We need to be careful about what happens inside these `sync` blocks other we might introduce deadlocks.

File: Source/Background/BackgroundActivityFactory.swift (hunk 1 above)
Comment (alexaubry) [reply Source/Background/BackgroundActivityFactory.swift]: I left it internal as I wanted to use them in tests initially, but they are only used by the factory in the end, so I will make them private. It is always called from another method, inside the sync queue.

**Comment by alexaubry** [comment]
@vytis We need changes in the upstream frameworks, yes. This is also an opportunity to see where we use background code and if we can improve it.
